lib/system/file_test.go

```diff
@@ -0,0 +1,142 @@
+package system
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+)
+
+// populates a test workspace (tmp) with a src containing several files
+func createSourceDir(t *testing.T) (tmp, src string, fcount int, err error) {
+	// create a workspace
+	tmp, err = ioutil.TempDir("", "robotest-test")
+	if err != nil {
+		t.Skipf("unable to create tempdir: %s", err)
+	}
+
+	// create content to move, several files to exercise the "all" portion
+	src = filepath.Join(tmp, "/src")
+	err = os.MkdirAll(src, 0750)
+	if err != nil {
+		t.Skipf("unable to create src dir: %s", err)
+	}
+
+	fcount = 3
+	for i := 0; i < fcount; i++ {
+		tmpfile, err := ioutil.TempFile(src, "file")
+		if err != nil {
+			t.Skipf("unable to create tempfile: %s", err)
+		}
+		defer tmpfile.Close()
+		content := []byte("file" + string(i))
+		if _, err := tmpfile.Write(content); err != nil {
+			t.Skipf("unable to write to %s: %s", tmpfile.Name(), err)
+		}
+	}
+	return tmp, src, fcount, nil
+}
+
+func TestCopyAllDirDstPresent(t *testing.T) {
+	tmp, src, fcount, err := createSourceDir(t)
+	defer os.RemoveAll(tmp)
+
+	dst := filepath.Join(tmp, "/dst")
+	err = os.MkdirAll(dst, 0750)
+	if err != nil {
+		t.Skipf("unable to create dst dir: %s", err)
+	}
+
+	var cnt uint
+	err = copyAll(src, dst, &cnt)
+	if err != nil {
+		t.Errorf("copy %q -> %q failed: %s", src, dst, err)
+	}
+	if int(cnt) != fcount {
+		t.Errorf("expected %v files to be copied, only %v copied", fcount, cnt)
+	}
+}
+
+func TestCopyAllDirDstAbsent(t *testing.T) {
+	tmp, src, fcount, err := createSourceDir(t)
+	defer os.RemoveAll(tmp)
+
+	// no destination directory
+	dst := filepath.Join(tmp, "/dst")
+
+	var cnt uint
+	err = copyAll(src, dst, &cnt)
+	if err != nil {
+		t.Errorf("copy %q -> %q failed: %s", src, dst, err)
+	}
+	if int(cnt) != fcount {
+		t.Errorf("expected %v files to be copied, only %v copied", fcount, cnt)
+	}
+	if _, err := os.Stat(dst); os.IsNotExist(err) {
+		t.Errorf("expected %v to be created", dst)
+	}
+}
+
+func TestCopyAllFileDstPresent(t *testing.T) {
+	// create a workspace
+	tmp, err := ioutil.TempDir("", "robotest-test")
```

Comment:
nit: you have `createSourceDir` but no `createSourceFile` which needlessly complicates the file-specific tests.

Reply:
My thought here was related to go's philosophy of "a little copying is better than a little dependency". 

https://go-proverbs.github.io/

When creating a directory with files in it, it was definitely enough work to abstract.  When creating a single file, it wasn't as clear to me that abstracting provided value, which is why I intentionally left it this way.

I trust your judgement though (since I'm learning what is appropriate in go) and will update!

Reply:
Addressed in https://github.com/gravitational/robotest/pull/174/commits/676d981935804a3dd1bdc6265ab92afc7a3c6214

Reply:
I'm in favor of tests that I can read and immediately understand what's being tested and how. Here's the method I'm trying to stick to when writing tests: http://xunitpatterns.com/Four%20Phase%20Test.html
For me this means a test should focus as much as possible on a thing being tested and contain as little irrelevant or setup code as possible. Using the current implementation as an example:
```go
	// setup
	tmp, src, fcount := createSourceFile(t)
	defer os.RemoveAll(tmp)
	// no destination directory
	dst := filepath.Join(tmp, "/dst")

	// exercise
	var cnt uint
	err := copyAll(src, dst, &cnt)

	// validate
	if err != nil {
		t.Errorf("copy %q -> %q failed: %s", src, dst, err)
	}
	if int(cnt) != fcount {
		t.Errorf("expected %v files to be copied, only %v copied", fcount, cnt)
	}
	if _, err := os.Stat(dst); os.IsNotExist(err) {
		t.Errorf("expected %v to be created", dst)
	}
```
Has nothing that would detract from understanding what the test case is about.

Reply:
Thanks for the link & the snippet.  I agree with your remark that

> a test should focus as much as possible on a thing being tested and contain as little irrelevant or setup code as possible

I typically lean towards fixtures for setup/teardown--to keep that out of the test case itself. pytest's parameterized fixtures are probably my favorite implementation (if a bit _magic_).  But I haven't reached that level of sophistication in go yet, as these are the first tests I've contributed in the language.  In lieu of fixtures, abstracting out common setup logic into a function does well enough.  Thanks for suggesting it.